pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/IndexCreationContext.java

```diff
@@ -0,0 +1,261 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.creator;
+
+import java.io.File;
+import java.util.Map;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.compression.ChunkCompressionType;
+import org.apache.pinot.segment.spi.index.creator.H3IndexConfig;
+import org.apache.pinot.spi.config.table.BloomFilterConfig;
+import org.apache.pinot.spi.config.table.FSTType;
+import org.apache.pinot.spi.data.FieldSpec;
+
+
+/**
+ * Provides parameters for constructing indexes via {@see IndexCreatorProvider}.
+ * The responsibility for ensuring that the correct parameters for a particular
+ * index type lies with the caller.
+ */
+public class IndexCreationContext {
+
+  public static final class Builder {
+    private File _indexDir;
+    private Map<String, Map<String, String>> _columnProperties;
+    private FSTType _fstType;
+    private H3IndexConfig _h3IndexConfig;
+    private int _lengthOfLongestEntry;
+    private int _maxNumberOfMultiValueElements;
+    private int _maxRowLengthInBytes;
+    private Object _sortedUniqueElementsArray;
+    private boolean _onHeap = false;
+    private ChunkCompressionType _chunkCompressionType;
+    public BloomFilterConfig _bloomFilterConfig;
+    private int _rangeIndexVersion;
+    private ColumnMetadata _columnMetadata;
```

Comment:
I feel we should not include `ColumnMetadata` within the `IndexCreationContext`. Conceptually column metadata should be available only after the indexes are created. Suggest putting the required properties such as field spec, min/max values as individual fields in the `IndexCreationContext`

Reply:
I agree, I was thinking about reverting the change to standardise on `ColumnMetadata`

Reply:
I thought about it and it needs to have `ColumnMetadata` because the range index needs to know the min and max values, because the index can't be built incrementally. I don't see the harm in bundling parameters from `SegmentColumnarIndexCreator` in `ColumnMetadata` just because min and max won't be known; it would only cause a problem if `SegmentColumnarIndexCreator` tried to create an index it didn't have enough to build.

Reply:
I will refactor this so there isn't one monolithic context

Reply:
Please see refactor here: https://github.com/apache/pinot/pull/7885/commits/38b54fccda1893aeeeccc7d0d198f4f48e62465c